jetty-alpn-agent 2.0.10 and akka-http 10.1.12

mkurz opened the pull request:
See https://webtide.com/jetty-alpn-java-8u252/

mkurz commented:
@raboof @jrudolph There is one failing test left now:
https://travis-ci.com/github/playframework/playframework/jobs/331459105#L1800-L1801

```
[info] EndpointIntegrationSpecificationSpec
[info] Endpoints should
[info]   respond with the highest supported HTTP protocol
[info]     + with Netty HTTP/1.1 (plaintext) (2 seconds, 191 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (2 seconds, 237 ms)
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (167 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (221 ms)
[error]     x with Akka HTTP HTTP/2 (encrypted) (196 ms)
[error]      http/1.1 != h2 (EndpointIntegrationSpecificationSpec.scala:26)
```

Is it possible this will fix itself when upgradin to akka-http 10.1.12 (akka/akka-http#3125)?

raboof commented:
> Is it possible this will fix itself when upgrading to akka-http 10.1.12 (akka/akka-http#3125)?

Hmm yes sounds like it might. Too bad we don't publish 10.1 snapshots currently, otherwise we could easily check this. Added https://github.com/akka/akka-http/pull/3140

mkurz commented:
@raboof How to retrieve the snapshot?

jrudolph commented:
It will probably still mean that HTTP/2 does not work on 8u252 where it did before (depending on whether the library which uses ALPN already has support for 8u252). However, it will prevent that TLS in general is broken.

mkurz commented:
@raboof @jrudolph This pull request is using the just published akka-http 10.1.11 snapshot now. However now another test fails: `play.it.http.AkkaIdleTimeoutSpec`

```
[info] AkkaIdleTimeoutSpec
[info] Play's idle timeout support should
[info]   + support null as an infinite timeout (1 second, 664 ms)
[info]   + support 'infinite' as an infinite timeout (5 seconds, 961 ms)
[error]   x support sub-second timeouts (4 seconds, 513 ms)
[error]    Expected: java.io.IOException. Got nothing (IdleTimeoutSpec.scala:124)
[error]   x support a separate timeout for https (11 seconds, 838 ms)
[error]    Expected: java.io.IOException. Got nothing (IdleTimeoutSpec.scala:141)
[error]   x support multi-second timeouts (28 seconds, 305 ms)
[error]    Expected: java.io.IOException. Got nothing (IdleTimeoutSpec.scala:151)
[info]   + not timeout for slow requests with a sub-second timeout (10 seconds, 74 ms)
[info]   + not timeout for slow requests with a multi-second timeout (7 seconds, 238 ms)
[info] Total for specification AkkaIdleTimeoutSpec
[info] Finished in 38 seconds, 510 ms
[info] 7 examples, 3 failures, 0 error
```
Did something change in regards to idle configs?

mkurz commented:
See https://travis-ci.com/github/playframework/playframework/jobs/331841648#L1582

mkurz commented:
Alright, I found the problem: Because of https://github.com/akka/akka-http/pull/2776 (which was backported to akka-http 10.1.12, for http2 the default idle timeout is infinite. And for some reason the `AkkaIdleTimeoutSpec` bind as http2... I think that's wrong, but I don't know yet why this is happening.

mkurz commented:
Alright, this should be it: I switched `IdleTimeoutSpec` to make use of `ServerEndpointRecipe`. Now the tests run with Http1 endpoints. I also added another test which runs akkahttp2 to makes sure the idletimeout is infinite in that case.

mkurz commented:
For the record, here is what the `IdleTimeoutSpec` results look like when running locally:

```
[info] IdleTimeoutSpec
[info] Play's idle timeout support should
[info]   support null as an infinite timeout
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (4 seconds, 42 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (4 seconds, 254 ms)
[info]     + with Netty HTTP/1.1 (plaintext) (5 seconds, 725 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (6 seconds, 276 ms)
[info]   support 'infinite' as an infinite timeout
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (4 seconds, 39 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (4 seconds, 245 ms)
[info]     + with Netty HTTP/1.1 (plaintext) (5 seconds, 724 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (6 seconds, 268 ms)
[info]   support sub-second timeouts
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (851 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (874 ms)
[info]     + with Netty HTTP/1.1 (plaintext) (3 seconds, 192 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (3 seconds, 239 ms)
[info]   support a separate timeout for https
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (1 second, 658 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (1 second, 51 ms)
[info]     + with Netty HTTP/1.1 (plaintext) (3 seconds, 819 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (3 seconds, 632 ms)
[info]   support multi-second timeouts
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (2 seconds, 971 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (2 seconds, 952 ms)
[info]     + with Netty HTTP/1.1 (plaintext) (7 seconds, 464 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (7 seconds, 591 ms)
[info]   not timeout for slow requests with a sub-second timeout
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (3 seconds, 133 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (3 seconds, 178 ms)
[info]     + with Netty HTTP/1.1 (plaintext) (5 seconds, 73 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (5 seconds, 196 ms)
[info]   not timeout for slow requests with a multi-second timeout
[info]     + with Akka HTTP HTTP/1.1 (plaintext) (7 seconds, 217 ms)
[info]     + with Akka HTTP HTTP/1.1 (encrypted) (7 seconds, 235 ms)
[info]     + with Netty HTTP/1.1 (plaintext) (9 seconds, 207 ms)
[info]     + with Netty HTTP/1.1 (encrypted) (9 seconds, 277 ms)
[info]   always be infinite when using akka-http HTTP/2
[info]     + with Akka HTTP HTTP/2 (plaintext) (10 seconds, 702 ms)
[info]     + with Akka HTTP HTTP/2 (encrypted) (10 seconds, 784 ms)
[info] Total for specification IdleTimeoutSpec
[info] Finished in 24 seconds, 696 ms
[info] 30 examples, 0 failure, 0 error
```

mkurz commented on core/play-integration-test/src/it/scala/play/it/http/IdleTimeoutSpec.scala:
```diff
@@ -85,96 +74,99 @@ trait IdleTimeoutSpec extends PlaySpecification with ServerIntegrationSpecificat
+        responses.length must_== 2
+        responses(0).status must_== 200
+        responses(1).status must_== 200
+      }
```
This last test here is new, to test https://github.com/akka/akka-http/pull/2776

mkurz commented:
Now build is failing because of missing akka-http api docs: https://github.com/akka/akka-http/issues/3143#issuecomment-627200356

ignasi35 commented:
Nice work @mkurz

ignasi35 commented:
Since you added `block-merge`, I'll leave it up to you to remove the label. ;-)

mkurz commented:
Thanks Ignasi, label removed :wink:
